libsel4vm/src/arch/arm/vgic/vgic_v3.c

```diff
@@ -0,0 +1,441 @@
+    gic_rdist->pidr4           = 0x44;     /* RO */
+
+    gic_rdist->cidr0           = 0x0D;     /* RO */
+    gic_rdist->cidr1           = 0xF0;     /* RO */
+    gic_rdist->cidr2           = 0x05;     /* RO */
+    gic_rdist->cidr3           = 0xB1;     /* RO */
+
+
+}
+
+
+int handle_vgic_maintenance(vm_vcpu_t *vcpu, int idx)
+{
+    /* STATE d) */
+    assert(vgic_dist);
+    struct gic_dist_map *gic_dist = vgic_priv_get_dist(vgic_dist);
+    vgic_t *vgic = vgic_device_get_vgic(vgic_dist);
+    assert(vgic);
+    vgic_vcpu_t *vgic_vcpu = get_vgic_vcpu(vgic, vcpu->vcpu_id);
+    assert(vgic_vcpu);
+    assert((idx >= 0) && (idx < ARRAY_SIZE(vgic_vcpu->lr_shadow)));
+    virq_handle_t *slot = &vgic_vcpu->lr_shadow[idx];
+    assert(*slot);
+    virq_handle_t lr_virq = *slot;
+    *slot = NULL;
+    /* Clear pending */
+    DIRQ("Maintenance IRQ %d\n", lr_virq->virq);
+    set_pending(gic_dist, lr_virq->virq, false, vcpu->vcpu_id);
+    virq_ack(vcpu, lr_virq);
+
+    /* Check the overflow list for pending IRQs */
+    struct virq_handle *virq = vgic_irq_dequeue(vgic, vcpu);
+    if (virq) {
+        return vgic_vcpu_load_list_reg(vgic, vcpu, idx, 1, virq);
+    }
+
+    return 0;
+}
+
+int vm_register_irq(vm_vcpu_t *vcpu, int irq, irq_ack_fn_t ack_fn, void *cookie)
+{
+    struct virq_handle *virq_data;
+    struct vgic *vgic;
+    int err;
+
+    vgic = vgic_device_get_vgic(vgic_dist);
+    assert(vgic);
+
+    virq_data = calloc(1, sizeof(*virq_data));
+    if (!virq_data) {
+        return -1;
+    }
+    virq_init(virq_data, irq, ack_fn, cookie);
+    err = virq_add(vcpu, vgic, virq_data);
+    if (err) {
+        free(virq_data);
+        return -1;
+    }
+    return 0;
+}
+
+
+int vm_inject_irq(vm_vcpu_t *vcpu, int irq)
+{
+    // vm->lock();
+
+    DIRQ("VM received IRQ %d\n", irq);
+
+    vgic_t *vgic = vgic_device_get_vgic(vgic_dist);
+    int err = vgic_dist_set_pending_irq(vgic, vcpu, irq);
+    if (!fault_handled(vcpu->vcpu_arch.fault) && fault_is_wfi(vcpu->vcpu_arch.fault)) {
+        ignore_fault(vcpu->vcpu_arch.fault);
+    }
+
+    // vm->unlock();
+
+    return err;
+}
+
+const struct vgic_dist_device dev_vgic_dist = {
+    .pstart = GIC_DIST_PADDR,
+    .size = 0x10000,
+    .priv = NULL,
+};
+
+
+const struct vgic_dist_device dev_vgic_redist = {
+    .pstart = GIC_REDIST_PADDR,
+    .size = 0xc0000,
+    .priv = NULL,
+};
+
+
+
+
+/*
+ * 1) completely virtual the distributor
+ * 2) remap vcpu to cpu. Full access
+ */
+int vm_install_vgic(vm_t *vm)
+{
+    vgic_t *vgic = calloc(1, sizeof(*vgic));
+    if (!vgic) {
+        ZF_LOGE("Unable to malloc memory for VGIC");
+        return -1;
+    }
+
+    vgic_reg_t *registers = calloc(1, sizeof(vgic_reg_t));
+    vgic->registers = registers;
+    if (!vgic->registers) {
+        free(vgic);
+        ZF_LOGE("Unable to malloc memory for VGIC");
+        return -1;
+    }
+
+    registers->dist = calloc(1, sizeof(struct gic_dist_map));
```

Comment:
I wonder, why do we allocate `dist` and `redist` separately and not make this a fixed part of `vgic_reg_t`, so we only have to allocate things once? That saves us 2 calls to `calloc()` and there would not be much difference because neither of these structs gets a special 4K alignment anyway.